### ISSUE TITLE
Multi Master Load Balance 

### ISSUE THREAD

**Issue by ccoughle** [saltstack/salt#48298]
### Description of Issue/Question
Our Multi Master set-up has a disparity in the load being placed on the servers.  
We have (depending on the time of year) between 1600 to 4000 minions connecting to a pair of salt masters.  While all minions do successfully connect to both masters.  The majority of the load is placed on a single master.  
It seems that the cause of this is that the network traffic for scheduled runs is almost entirely directed to a single master.  This results in an approximately 10 fold increase in traffic to the master handling the traffic in comparison to the non stressed master.  
So far I have been unable to isolate what mechanism is determining which master ends up with the heaver load.  A few tests I have run and the results are summarized below:  

1:  Shutting down the heavy load master does cause the load to shift to the other master.  Bringing the heavy load master back up causes the load to flip back to it.  
2:  I have altered the other of our two masters in the minion.conf.  This has had no effect on which master received the heavier load.   
3:  The two masters are in different data-centers.  So, I have spun up additional testing masters in both data-centers and there is no correlation between data center and increased load.  It seems random which master is chosen and once chosen it seems to be stuck to that master.  
4:  In the off chance that the random_master setting in the minion conf would affect this at all (even though it is supposed to be only applicable to failover mode) I tried with and without random_master set to true,  This had no effect.  

Please let me know what additional information might be helpful.  We have had this problem in the past and it has gotten cleared up with updates and then returned again.  Generally we can manage the load with a single server.  However, when our usage ballons during a semester and approach more than 4000 minions we have encounter issues that have put things into a bad state.  

### Setup

Minion.conf changes from default:

master:
  - salt-01.department.university.edu
  - salt-02.department.university.edu

hash_type: sha256

rejected_retry: True

### Versions Report
Salt Version:
           Salt: 2018.3.1

Dependency Versions:
           cffi: Not Installed
       cherrypy: unknown
       dateutil: 1.5
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.1
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.5 (default, Apr 11 2018, 07:36:10)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: centos 7.5.1804 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-862.3.3.el7.x86_64
         system: Linux
        version: CentOS Linux 7.5.1804 Core

**Comment by garethgreenaway** [comment]
@ccoughle Thanks for the report.  There are a couple different minion configuration:
https://docs.saltstack.com/en/2017.7/ref/configuration/minion.html#master-type
which should be set to `failover`
and
https://docs.saltstack.com/en/2017.7/ref/configuration/minion.html#master-shuffle
which should be set to True
and finally (which I believe you're have configured but might not be being used if you don't have the failover setting correct):
https://docs.saltstack.com/en/2017.7/ref/configuration/minion.html#random-master
which should be set to True.

**Comment by ccoughle** [comment]
Hello, 

I did run the machines in a failover mode as a test.  However, the outcome of that was not what we needed as the minions only make a connection to one of the masters at a time, while in the standard (default master_type: str) set-up there are connections open to both masters at the same time.  

Regarding the master_shuffle and random_master settings, I just included the one in case there was an bug wrt the master connection for the default master_type without the random setting that resulted in the establishment of a primary master regardless of what the confs directed.  (I knew this was a loooong shot, but wanted to try every thing that I could before submitting a report)  

Lastly, according to:  
https://github.com/saltstack/salt/issues/40532
random_master and master_shuffle are basically redundant with masste_shuffle being depreciated.  

Basically, it seems as if the minions when in a multi-master set-up are connecting to both masters as expected, however, all network traffic (pillar encryption and file transfer) is only coming from one of the masters.  

Thanks

Casey

**Comment by ccoughle** [comment]
Oh, one other thing I should mention is that a manual salt-call state.apply from the minions does randomly select which master to connect to.  I know that the scheduled runs and the manual salt-calls have different entry points into the system  (can't remember where we wound that in the python files exactly), but I figured that would be a good data point to pass along.

-Casey

**Comment by jacksontj** [comment]
I actually see this same behavior as well. In my setup I have each minion connecting with 2 masters. The issue seems to be that the second one in the list seemingly gets all the load. From my initial investigation it seems to be a race-condition with how the minion is started -- and the load is actually from the scheduled highstates, not regular commands etc.. The MultiMinion actually starts up 2 schedulers and for whatever reason the second master in the list gets all the load (you can actually see in the logs entries about it not starting jobs because one is running). This really needs to change on the minion-- what seems right to me is to have the minion create 1 scheduler and RR the master *per schedule run* such that all executions within that schedule job (highstate in this example) all go to the same master, but that way they'll random across things.

**Comment by ccoughle** [comment]
My original testing did show consistency of which master grabbed all the load.  However, I could not tie it to any particular setting.  
Changing the master order in the minion and master confs made no difference.  
I added masters to the pool to see if I could tie it to the hardware pool and/or network they were on.  Again, no difference.  
Currently, because of issues with scheduled runs in general,  (schedule runs are failing because of python issues related to our group file on the minions being over some undefined size) I have had to switch all my minions over to cron.job scheduled runs with salt-call.  This gets around the scheduled run issues as well as the master load as the salt-calls from the minion actually distribute the load correctly.
